pkg/types/contract_reader.go

```diff
@@ -65,6 +65,9 @@ type ContractReader interface {
 	// QueryKey provides fetching chain agnostic events (Sequence) with general querying capability.
 	QueryKey(ctx context.Context, contract BoundContract, filter query.KeyFilter, limitAndSort query.LimitAndSort, sequenceDataType any) ([]Sequence, error)
 
+	// QueryKeys provides fetching chain agnostic events of different types (SequenceWithKey) with general querying capability.
+	QueryKeys(ctx context.Context, contract BoundContract, filters []query.KeyFilter, limitAndSort query.LimitAndSort, sequenceDataType any) ([]SequenceWithKey, error)
```

Comment:
The problem with map[string]Sequence is it would lose the order, the key thing is that the order across the various event types needs to be maintained.

no strong feelings on using iter, however the result isn't streamed so no strong advantage to using iter either as far as I can see, and I think it would add a bit of complexity to the loop implementation.  On balance my pref would be for []SequenceWithKey

Reply:
Streaming is not the only reason to use an `iter.Seq`, and this API is a layer above the GRPC protos. The advantage is that we can express what we need naturally and directly without declaring extra hyper-specific types. Another similar approach could be a generic entry/pair/etc.:
```go
type Entry[K comparable, V any] struct {
	K K
	V V
}
```
```suggestion
	QueryKeys(ctx context.Context, contract BoundContract, filters []query.KeyFilter, limitAndSort query.LimitAndSort, sequenceDataType any) ([]Entry[string, Sequence], error)
```
Could also consider documenting via a type:
```go
// SortedEntries is a slice of [Entry]s sorted by ascending keys.
type SortedEntries[K comparable, V any] []Entry[K, V]
```
```suggestion
	QueryKeys(ctx context.Context, contract BoundContract, filters []query.KeyFilter, limitAndSort query.LimitAndSort, sequenceDataType any) (SortedEntries[string, Sequence], error)
```

Reply:
The name Key I think is a bit confusing here, however to be consistent with the existing API I did not rename it, it is not a comparable key but actually the event type.   I wasn't party to the original design discussion on contract reader but I assume there is a good reason for using that name related to keeping it  generic across chain implementations, so in practice they are not comparable.

Reply:
If they are not `comparable`, then in what order are they sorted?

Reply:
```
// ----------------------- Solidity types -----------------------
event Transfer(address indexed from, address indexed to, uint256 value);
event Approval(address indexed owner, address indexed spender, uint256 value);
event SomeOtherEvent(a uint32 indexed, b bool, c string)

// ----------------------- Go types -----------------------
type Transfer struct{
  From, To common.Address
  Value *big.Int
}

type Approval struct{
  Owner, Spender common.Address
  Value *big.Int
}

type SomeOtherEvent struct{
  A  *big.Int
  B  bool
  C  string
}

// ----------------------- Querying -----------------------

filters := []ContractKeyFilter {
{
 KeyFilter: {Key: "transfer", ... Filter: someFilter}
 Contract: usdcContract,
 SequenceDataType: &Transfer{}
},
{
 KeyFilter: {Key: "Approval", ... Filter: someOtherFilter}
 Contract: usdcContract,
 SequenceDataType: &Approval{}
},
{
 KeyFilter: {Key: "transfer", ... Filter: anotherFilter}
 Contract: someOtherContract,
 SequenceDataType: &SomeOtherEvent{}
},
}
QueryKeys(ctx, filters, query.NewLimitAndSort(query.Limit{}, query.NewSortBySequence(query.Asc)))

// ----------------------- Result-----------------------

If we query in block range 100-150, where these 3 events were emitted:
Transfer block: 123 logIndex = 1 
Approval block: 123 logIndex = 3
SomeOtherEvent block: 100 logIndex = 4


Result should be ASC based on block-logIndex (sequence) ordering stored in Cursor: 
[]SequenceWithKey{
{Key: "SomeOtherEvent", Sequence:{ Cursor: "100-4", Data: &SequenceDataType{...}},
{Key: "Transfer", Sequence:{ Cursor: "123-1", Data: &Transfer{...}},
{Key: "Approval", Sequence:{ Cursor: "123-3", Data: &Approval{...}}
}
```

Reply:
Events should be sorted by their onchain sequence which is different for every chain, above I highlighted an EVM example where sorting is done based on block-logIndex.

Reply:
The problem that we are trying to solve is caller trying to fetch different things with QueryKey and later sorting them in the order that they happened onchain. We don't have anything comparable that we can expose as the Cursor is just a string. One solution is to implement a CompareCursor on the relayer to compare which sequence comes before which, another is to implement a whole sorting function and the third solution is to give back multiple different types already sorted which is done in this PR

Reply:
> 1. Please document this on the interface.
CC @ettec 

> 2. How do you expect users to interact with the returned data?
The caller has to keep track of which type belongs to which key. To parse the response the caller would have to loop through the results and cast the data to concrete types based on the associated key.

Reply:
At the GRPC level, I think a response with a `repeated` field like `SequenceWithKey`(`KeyedSequence`?) makes sense, and the `iter.Seq2` could be backed by that generated proto type.

Reply:
commons on go 1.22, iters are a go 1.23 feature.    Don't have the bandwidth to upgrade common at the moment, it may be straightforward to upgrade (?). but I'm guessing there may be some unseen side effect?

Reply:
Looks like nbd in this case: 
- #952